File: backend/app/schemas/paper.py
"""Paper Pydantic schemas."""

from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import TagRef


class CategoryRef(BaseModel):
    id: str
    name: str
    source: str

    model_config = {"from_attributes": True}


class AuthorRef(BaseModel):
    id: str
    name: str
    affiliation: Optional[str] = None
    author_order: int = 0

    model_config = {"from_attributes": True}


class PaperVersionRef(BaseModel):
    id: str
    title: str
    source: str
    url: str

    model_config = {"from_attributes": True}


class PaperSummary(BaseModel):
    id: str
    title: str
    source: str
    source_id: str
    venue: Optional[str] = None
    published_date: Optional[date] = None
    url: str
    pdf_url: Optional[str] = None
    code_url: Optional[str] = None
    crawled_at: datetime
    updated_at: datetime
    categories: list[CategoryRef] = Field(default_factory=list)
    tags: list[TagRef] = Field(default_factory=list)
    author_names: list[str] = Field(default_factory=list)
    summary_cn: Optional[str] = None
    summary_en: Optional[str] = None

    model_config = {"from_attributes": True}


class PaperDetail(PaperSummary):
    abstract: Optional[str] = None
    authors: list[AuthorRef] = Field(default_factory=list)
    merged_into_id: Optional[str] = None
    versions: list[PaperVersionRef] = Field(default_factory=list)


class PaperTagUpdate(BaseModel):
    tag_ids: list[str]


class PaperTagResponse(BaseModel):
    paper_id: str
    tags: list[TagRef]

File: frontend/src/types/paper.ts
import type { CategoryRef } from './common';
import type { TagRef } from './tag';

// ---- Author Ref ----
export interface AuthorRef {
  id: string;
  name: string;
  affiliation: string | null;
  author_order: number;
}

// ---- Paper Summary (list item) ----
export interface PaperSummary {
  id: string;
  title: string;
  source: string;
  source_id: string;
  venue: string | null;
  published_date: string | null;
  url: string;
  pdf_url: string | null;
  code_url: string | null;
  crawled_at: string;
  updated_at: string;
  categories: CategoryRef[];
  tags: TagRef[];
  author_names: string[];
  summary_cn: string | null;
  summary_en: string | null;
}

// ---- Paper Detail (full detail) ----
export interface PaperDetail extends PaperSummary {
  abstract: string | null;
  authors: AuthorRef[];
  merged_into_id: string | null;
  versions: PaperVersionRef[];
}

// ---- Paper Version Ref ----
export interface PaperVersionRef {
  id: string;
  title: string;
  source: string;
  url: string;
}

// ---- Author Paper Ref ----
export interface AuthorPaperRef {
  id: string;
  title: string;
  published_date: string | null;
  source: string;
  author_order: number;
}

// ---- Author With Papers ----
export interface AuthorWithPapers {
  id: string;
  name: string;
  affiliation: string | null;
  papers: AuthorPaperRef[];
}

// ---- Author Summary (list item) ----
export interface AuthorSummary {
  id: string;
  name: string;
  affiliation: string | null;
  paper_count: number;
}

// ---- Paper Query Params ----
export interface PaperQueryParams {
  page?: number;
  page_size?: number;
  source?: string[];
  category?: string[];
  date_from?: string;
  date_to?: string;
  q?: string;
  tag_id?: string[];
  sort?: string;
  venue?: string;
}
